### ISSUE TITLE
UI Scale Issue with New Renderer

### ISSUE THREAD

**Issue by Hexafish** [PathOfBuildingCommunity/PathOfBuilding#7414]
I have a 2560x1440 screen. However, I usually don't like my applications such as Firefox as well as the native windows apps being tiny, so I scale my applications to 150% in Windows.

Unfortunately, Path of Building respects this as well, and its the one application I prefer at 100% so I can utilize more of the screen. Previously with the old rendering engine, I was able to fix this by right clicking the app, selecting Properties -> Compatibility -> Change high DPI settings -> Checking Override high DPI scaling behavior -> Application, and PoB would render as though the application scale was 100%.

With this new engine, that doesn't work and I need to change my windows scale to 100% every time which is not convenient as it messes with Firefox and also makes interacting with anything weird as my mouse DPI is not affected by resolution scales, so the mouse moves slower.

**Comment by zao** [comment]
This is a fun niche use case that I didn't consider with the new runtime.

## Context
In the original runtime, the program didn't know anything about DPI and as such, Windows defaulted to scaling it up from the outside. As you noted you could lie to Windows and tell it that the program could handle DPI itself and thus rendered naively at 100% scale.

In the new runtime, the program has a DPI-aware window out of the box and as such, altering the choice in Windows between System and Application has no effect, as System understands the program's declaration and transfers control to the program.

In the program, as we're now responsible for scaling the graphics and input ourselves we set up a low resolution render target and scale that up with the DPI scaling factor we obtain from Windows. The funny thing is that if we were to do nothing, we'd end up rendering at 100% scale like you would expect, but unlike what most user that intentionally configure a scale in Windows would expect.

## Workaround
I don't have any immediate long-term solution for you unfortunately. This stretching-on-our-own functionality was to keep usability until a future date when we've made the UI more scalable in itself and can allow it to draw at full resolution with the individual controls at scaled dimensions.

The best I can offer you now is a hack which will be undone whenever there's an PoB update as it modifies the Lua code.
You would be replacing `RenderInit()` in `src/Launch.lua` with `RenderInit("DPI_AWARE")`, which tricks the runtime into thinking that the Lua code can draw things at scale, which it can't and as such will draw at 100%.

https://github.com/PathOfBuildingCommunity/PathOfBuilding/blob/4f7ffcbe9fbd08fe3d085afe81489dfabf152645/src/Launch.lua#L64-L68

If you opt to do that, don't forget to actually update PoB occasionally and reapply the fix afterwards.

It's possible that we might be able to do some other sort of hidden knob to steer the scaling factor for power users, but I haven't thought that through yet.

**Comment by Hexafish** [comment]
That seems like a surprisingly simple fix for something that I assumed would be slightly more complicated to fix given the explanation.

**Comment by istarogh** [comment]
Thanks for the workaround instructions.  I, uh, tried making that edit to the Launch.lua file in my PoB directory (`%AppData%/Path of Building Community`), but that prevented PoB from loading at all, even after I changed it back, until I redownloaded Launch.lua from GitHub.  I'm guessing you meant not "just change the local file" but rather something involving downloading all the code, making the change, and figuring out how to do the build myself?

In any case, I wanted to add my voice, a "+1" so to speak, to the effect that maybe this isn't quite _that_ niche an issue.  I run Windows 10 at a custom scaling factor of 125% to ease eye strain.  Which, if I correctly recall the process I went through many years ago and why I did so, is different from a non-custom scaling factor of 125% in that it prevents legacy programs from exhibiting that characteristic "DPI problem" fuzzy text.  With the exception, now, of PoB.  That is, before today it looked like this:
![oldcrisp](https://github.com/PathOfBuildingCommunity/PathOfBuilding/assets/83389704/20731b9d-41e5-42c0-97d6-905915cf1677)
and now it looks like this:
![newfuzzy](https://github.com/PathOfBuildingCommunity/PathOfBuilding/assets/83389704/d1e7e82d-d174-4d04-bffd-9ef40b95e549)

(Straight lines, for example, used to be crisp single-pixel lines, and now are accompanied by ghosting.)

**Comment by chx** [comment]
To add a different voice, PoB was very near unusable before for me at 200% scaling on a 55" HDTV so I used the compatibility fix but that made it almost impossible to read , it was so tiny. I can live with a little blurriness if that's the price for otherwise usable PoB.

**Comment by Terroosh** [comment]
I don't know if niche would be correct here. Myself and all my friends play using various resolutions all with scaling enabled. In fact many gaming laptops come with some sort of scaling % enabled by default. In any case, all of us have a hard time using PoB because of how tiny the text size is so hopefully at some point PoB will be able to render at above 100% without the blurry text/smearing issue.

**Comment by zao** [comment]
Varying display densities and preferences are indeed way more common these days, I didn't intend to diminish the impact for any group.

I've wanted to do crisper UI for a long while as I daily drive dense displays at 150%-200%.
On custom vs. non-custom scale factor, I don't know what Windows does differently there.

There's some existing body of research out there on content-aware scaling in emulators and even Windows itself, but researching and implementing something like that also takes time.

**Comment by Silverun** [comment]
After 2.41 update my PoB looked like 800x600 on fullHD monitor. I think I have 125 scaling in Windows 10. This workaround helped.

**Comment by Sheodar** [comment]
Bump? Can we do a bump?

**Comment by missingdays** [comment]
I have a 4k monitor and the scaling setup in the OS settings are not respected

![image](https://github.com/user-attachments/assets/944988b7-8a49-4e72-a50f-328c685047ab)

Workaround with `RenderInit("DPI_AWARE")` also didn't help

**Comment by zao** [comment]
@missingdays As this screenshot seems to be from some Linux distribution, I'm not certain which flavour of PoB you're running.
Is it the real Windows version we ship running under Wine or a third party package like the one from Arch AUR or the Flatpak packages?

For our real version, we declare at startup that we are DPI-aware and thus want to handle scaling ourselves.
We query the DPI scaling value from the OS and draw to a proportionally lower resolution render target which is upscaled to the desired scale for presentation.

`RenderInit("DPI_AWARE")` as a workaround inhibits this mechanism in our runtime, drawing the UI at 1:1 to underlying pixels. The real purpose of this flag is to allow the Lua-side UI to query the scale factor, pick font sizes and place/size UI elements accordingly to get a crisp high-resolution UI. It happens that it's also a workaround for people who _have_ functioning DPI scaling in the OS but wish to keep PoB at 1:1.

You might have to dig into your environment and Wine yourself and see if you can get it to forward or honor DPI scaling, we have extremely limited bandwidth for addressing interactions like this when we use the proper Windows API functions already.

**Comment by missingdays** [comment]
@zao I installed Path of Building from Lutris, here https://lutris.net/games/path-of-building/

> https://lutris.net/games/path-of-building/

Of course, I understand how low of a priority such issues must be. I'm just adding a datapoint to the issue that seemed relevant

**Comment by missingdays** [comment]
What helped me was going to Wine config in Lutris
![image](https://github.com/user-attachments/assets/2fd8173f-ab36-44c0-a44d-0165ab885f4e)
and in graphics changing the `screen resolution` value
![image](https://github.com/user-attachments/assets/8e4b6a63-9fb6-4286-bece-e699931bf3a9)
After that, Path of Building is rendered in a more readable resolution. It looks a bit blocky, but much more usable

**Comment by Silverun** [comment]
Hey, I have ran troubleshooter for the app and the issue has been fixed! 
I have 125 % scaling in system. PoB was blurry, OP fix helped but you needed to apply it every time after update.

Go to PoB.exe properties - Compatability - Change high DPS settings - Override high DPI setting - Application

![image](https://github.com/user-attachments/assets/091a02db-81b9-4973-84e6-6231b7967341)

**Comment by Sheodar** [comment]
Yeah, that's all cool, but number and words in this case will be too small for eyes. So will be cool for some render magic for PoB could work with win scaling.
